lib/client/ca_export.go

```diff
@@ -0,0 +1,289 @@
+// To do so, set CompatVersion to "1.0".
+// No other CompatVersion value is accepted.
+type ExportAuthoritiesRequest struct {
+	AuthType                   string
+	ExportAuthorityFingerprint string
+	ExportPrivateKeys          bool
+	CompatVersion              string
+}
+
+// ExportAuthorities returns the list of authorities in OpenSSH compatible formats as a string.
+// If the ExportAuthoritiesRequest.AuthType is present only prints keys for CAs of this type,
+// otherwise returns all keys concatenated.
+//
+// Exporting using "tls*", "database", "windows" AuthType:
+// Returns the certificate authority public key to be used by systems that rely on TLS.
+// The format can be PEM or DER depending on the target.
+//
+// Exporting using "user" AuthType:
+// Returns the certificate authority public key exported as a single
+// line that can be placed in ~/.ssh/authorized_keys file. The format adheres to the
+// man sshd (8) authorized_keys format, a space-separated list of: options, keytype,
+// base64-encoded key, comment.
+// For example:
+// > cert-authority AAA... type=user&clustername=cluster-a
+// URL encoding is used to pass the CA type and cluster name into the comment field.
+//
+// Exporting using "host" AuthType:
+// Returns the certificate authority public key exported as a single line
+// that can be placed in ~/.ssh/authorized_hosts. The format adheres to the man sshd (8)
+// authorized_hosts format, a space-separated list of: marker, hosts, key, and comment.
+// For example:
+// > @cert-authority *.cluster-a ssh-rsa AAA... type=host
+// URL encoding is used to pass the CA type and allowed logins into the comment field.
+func ExportAuthorities(ctx context.Context, client auth.ClientI, req ExportAuthoritiesRequest) (string, error) {
+	var typesToExport []types.CertAuthType
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.HostCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "db":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.DatabaseCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user-der", "windows":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         true,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	}
+
+	// If no AuthType is given, export both HostCA and UserCA.
```

Comment:
Is everything below this point either an SSH authorized key or authorized host? Could we say that in a comment?

Reply:
Yes
Adding a comment

Reply:
Much better, thanks.